src/psyclone/tests/psyir/transformations/inline_trans_test.py

```diff
@@ -772,6 +772,172 @@ def test_apply_multi_function(fortran_reader, fortran_writer, tmpdir):
+    inlined is in a different container and is within a module (so
+    a use statement is required).
+
+    '''
+    code = (
+        f"module inline_mod\n"
+        f"contains\n"
+        f"  subroutine sub(x)\n"
+        f"    real, intent(inout) :: x\n"
+        f"    x = 2.0*x\n"
+        f"  end subroutine sub\n"
+        f"end module inline_mod\n"
+        f"module test_mod\n"
+        f"{use1}"
+        f"contains\n"
+        f"  subroutine run_it()\n"
+        f"    {use2}"
+        f"    real :: a\n"
+        f"    call sub(a)\n"
+        f"  end subroutine run_it\n"
+        f"end module test_mod\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    inline_trans.apply(routine)
+    output = fortran_writer(psyir)
+    assert (
+        "    real :: a\n\n"
+        "    a = 2.0 * a\n\n"
+        "  end subroutine run_it" in output)
+    assert Compile(tmpdir).string_compiles(output)
+
+
+def test_apply_container_subroutine_error(fortran_reader):
+    '''Test the apply method does not find the Routine if the Call is
+    imported but the Routine is not in a Container.
+
+    '''
+    code = (
+        "module test_mod\n"
+        "use inline_mod, only : sub\n"
+        "contains\n"
+        "  subroutine run_it()\n"
+        "    real :: a\n"
+        "    call sub(a)\n"
+        "  end subroutine run_it\n"
+        "end module test_mod\n"
+        "subroutine sub(x)\n"
+        "  real, intent(inout) :: x\n"
+        "  x = 2.0*x\n"
+        "end subroutine sub\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    with pytest.raises(TransformationError) as info:
+        inline_trans.apply(routine)
+    assert ("Failed to find the source for routine 'sub' and therefore "
+            "cannot inline it." in str(info.value))
+
+
+def test_apply_call_container_error(fortran_reader):
+    '''Test the apply method does not find the Routine if the Call is
+    not imported but the Routine is in a Container.
+
+    '''
+    code = (
+        "subroutine run_it()\n"
+        "  real :: a\n"
+        "  call sub(a)\n"
+        "end subroutine run_it\n"
+        "module inline_mod\n"
+        "contains\n"
+        "subroutine sub(x)\n"
+        "  real, intent(inout) :: x\n"
+        "  x = 2.0*x\n"
+        "end subroutine sub\n"
+        "end module inline_mod\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    with pytest.raises(TransformationError) as info:
+        inline_trans.apply(routine)
+    assert ("Failed to find the source for routine 'sub' and therefore "
+            "cannot inline it." in str(info.value))
+
+
+def test_apply_indirect_use_error(fortran_reader):
+    '''Test the apply method does not find the Routine if the Call is
+    imported but the Routine is not found in the specified Container.
```

Comment:
Ah, is this what you were saying in that comment I complained about? It doesn't feel as though this is difficult to fix. Can `_find_routine_in_container` be called recursively?

Reply:
OK, I've added that.

Reply:
Thanks.